### PR TITLE
MBS-13449: Reimplement transactional cache without database locks

### PR DESCRIPTION
**Pull request by mwiencek**
# Problem

The current implementation for MBS-7241 causes deadlocks and performance issues due to lock contention (MBS-11345, MBS-12314).

# Solution

Implement a different solution that blocks cache additions using our Redis store to manage "cachelock" keys.

The new implementation is documented in `Data::Role::EntityCache`. Please review that to make sure it seems sound.

# Testing

The existing test for MBS-7241 in `GIDEntityCache` passes. However, this hasn't been tested at scale. Complicating things, it also needs to be deployed to production and beta at the same time.

### CODE REVIEW

**Comment by mwiencek** [comment]
BTW, there is at least one minor race condition here:

 * Request A reads artist ID=1 from the database. It's not in the cache, and there is no cachelock key.
 * Request B starts before request A adds artist ID=1 to the cache. It updates or deletes artist ID=1 and fully commits.
 * Request A finally adds (the old version of) artist ID=1 to the cache.

However, the window of time for this to happen is extremely narrow, so is probably more acceptable than the current situation. I'll try to think of ways to mitigate this, too.

File: lib/MusicBrainz/Server/Data/Role/EntityCache.pm [hunk 1]
```diff
@@ -135,43 +150,32 @@ sub _delete_from_cache {
...
+                $c->store->delete_multi(@cachelock_keys);
+            });
         }
     }
 
-    my $cache_prefix = $self->_type . ':';
     my @keys = map { $cache_prefix . $_ } @ids;
-
-    my $cache = $self->c->cache($self->_type);
+    my $cache = $c->cache($self->_type);
```
Comment (reosarevok) [inline lib/MusicBrainz/Server/Data/Role/EntityCache.pm:176]: What's the point of this being a variable if you don't use it anymore in 178?

**Review by yvanzo — COMMENTED** [review]
LGTMBDNT. The comments to the code are very explanatory. I don’t know how we could reproduce potential race conditions in CI tests, so it’s fine to skip for now.

Is there anything special to be done when deploying on both instances? Like clearing some caches? Please document these extra steps if any somewhere.

> BTW, there is at least one minor race condition here:
> 
>     * Request A reads artist ID=1 from the database. It's not in the cache, and there is no cachelock key.
> 
>     * Request B starts before request A adds artist ID=1 to the cache. It updates or deletes artist ID=1 and fully commits.
> 
>     * Request A finally adds (the old version of) artist ID=1 to the cache.
> 
> However, the window of time for this to happen is extremely narrow, so is probably more acceptable than the current situation. I'll try to think of ways to mitigate this, too.

Would it work to require request A to acquire the cachelock before loading the entity from the database?

**Comment by mwiencek** [comment]
@yvanzo Thanks for the review; actually I think I found a way to reproduce the race condition in the tests, and was working on a fix, but haven't pushed the changes yet.

> Is there anything special to be done when deploying on both instances? Like clearing some caches? Please document these extra steps if any somewhere.

Right, I will think about this more and add some documentation.

> Would it work to require request A to acquire the cachelock before loading the entity from the database?

It would work if it was an exclusive lock, I think, though it's what I'm trying to move away from to avoid lock contention: the "deleting" transaction doesn't actually read the cachelock key at all, only sets it in order to temporarily prevent any "reading" transaction from adding to the cache.

There is one thing that I think would help, and that is https://redis.io/docs/interact/transactions/#watch-explained which we could use to `WATCH` the cachelock key; so when "request A finally adds (the old version of) artist ID=1 to the cache," it would not apply, because it would see that the cachelock key was modified in the meantime.

The only problem with this is that the current implementation here stores the cachelock keys in our KeyDB *store* rather than the *cache*.  Watching a key in the store will have no effect on the cache.  So we'd have to move the cachelock keys to the cache, which seems iffy, since it means they can *theoretically* be evicted before we're done with them.  However, perhaps this isn't the issue it seems, since we use the LRU eviction policy and it should be very unlikely that it's evicted in the time span we're talking about.  It would be interesting if we had some stats regarding "average age of keys when they're evicted."

File: lib/MusicBrainz/Server/Data/Role/EntityCache.pm (hunk 1 above)
Comment (mwiencek) [reply lib/MusicBrainz/Server/Data/Role/EntityCache.pm:176]: This might be something I cleaned up locally but didn't push yet. Will make sure it gets fixed if not. 👍

**Comment by mwiencek** [comment]
I renamed `cachelock` to `recently_invalidated` to make it clear that it's not an actual lock. (Let me know if you prefer the previous name, or have a better suggestion.)

I also made some changes which should help resolve the race condition described above:
 * Instead of checking for a `recently_invalidated` key before we add something to the cache, we check it immediately *after* we add something to the cache, and just delete the added entry right away if it was in fact recently-invalidated. (We could also check it before, but I feel that there may be more of a downside to adding too many cache lookups to handle an already-rare occurrence.)
 * The `recently_invalidated` keys are no longer deleted after the deleting transaction ends; this was a mistake because there may still be stale inserting transactions open after the deleting one commits. Instead, we always leave them to expire after the max request time has elapsed.

A test has been added for the scenario described in https://github.com/metabrainz/musicbrainz-server/pull/3130#issuecomment-1863226174, and another commit has been added to follow-up on an issue I found for https://github.com/metabrainz/musicbrainz-server/pull/2751#discussion_r1417976789.

**Comment by mwiencek** [comment]
> Is there anything special to be done when deploying on both instances? Like clearing some caches? Please document these extra steps if any somewhere.

I don't think anything special needs to be done, actually, since there is nothing currently in the cache that would conflict with this PR. It's only making use of some new keys being added.

After deploying this to beta, there is a minor chance that the type of issue described by MBS-7241 can arise -- either a deleted entity remaining in the cache, or the cache returning an outdated version of an entity -- until it's also deployed to production.

**Comment by yvanzo** [comment]
Do you now intend to deploy it with the usual `beta` rather than to deploy it to both `beta` and `production` at the same time?
Should it be tracked with a separate ticket?

**Comment by mwiencek** [comment]
@brainzbot test this please

> Do you now intend to deploy it with the usual beta rather than to deploy it to both beta and production at the same time?
Should it be tracked with a separate ticket?

Added MBS-13449. I think deploying it to production still makes sense, since if any cache issues arise, it might otherwise be unclear if it's due to a bug in the code, or due to a lack of symmetry between the code on production and beta.

**Comment by mwiencek** [comment]
@brainzbot test this please

**Comment by mwiencek** [comment]
@brainzbot test this please

**Comment by mwiencek** [comment]
@brainzbot test this please

**Comment by mwiencek** [comment]
@brainzbot test this please

**Comment by mwiencek** [comment]
https://ci.metabrainz.org/job/musicbrainz-server/10292/ passed (which was accidentally based on master), and there is another test run queued based on production. (But I think the above run is sufficient to show that tests work now.)

File: t/lib/t/MusicBrainz/DataStore/RedisMulti.pm [hunk 2]
```diff
@@ -23,10 +23,10 @@ stores. Consistency among all the stores is checked after each operation.
 
 # Initialize tests
 my $args1 = DBDefs->DATASTORE_REDIS_ARGS;
-$args1->{database} = $args1->{test_database};
+$args1->{database} = DBDefs->REDIS_TEST_DATABASE;
 
 my $args2 = DBDefs->DATASTORE_REDIS_ARGS;
-$args2->{database} = $args2->{test_database} + 1;
+$args2->{database} = DBDefs->REDIS_TEST_DATABASE + 1;
```
Comment (reosarevok) [inline t/lib/t/MusicBrainz/DataStore/RedisMulti.pm:29]: Given this here, should the docs in DBDefs say that both `REDIS_TEST_DATABASE` and `REDIS_TEST_DATABASE + 1` will be completely erased on each test run? Or should we define `REDIS_SECOND_TEST_DATABASE`? You know someone will use 0 for test and 1 for production some day.

File: lib/MusicBrainz/Server/Data/Release.pm [hunk 3]
```diff
@@ -1714,6 +1714,8 @@ sub update_amazon_asin {
             { id => $release->id },
         );
     }
+
+    $self->_delete_from_cache($release_id);
```
Comment (reosarevok) [inline lib/MusicBrainz/Server/Data/Release.pm:1718]: Shouldn't this technically be its own commit? :p

File: t/lib/t/MusicBrainz/DataStore/RedisMulti.pm (hunk 2 above)
Comment (mwiencek) [reply t/lib/t/MusicBrainz/DataStore/RedisMulti.pm:29]: Yeah, I think we should just mention that any Redis database numbers greater than or equal to `REDIS_TEST_DATABASE` may be used for tests.

File: lib/MusicBrainz/Server/Data/Release.pm (hunk 3 above)
Comment (mwiencek) [reply lib/MusicBrainz/Server/Data/Release.pm:1718]: Good point, moved it to a separate commit.

**Comment by mwiencek** [comment]
(Ignore the pushed diffs: I've accidentally rebased on master several times when this PR is based on production, and had to revert the changes.)

**Comment by mwiencek** [comment]
@brainzbot test this please

**Comment by mwiencek** [comment]
@brainzbot test this please

**Comment by mwiencek** [comment]
@brainzbot test this please

**Comment by mwiencek** [comment]
@brainzbot test this please

File: t/lib/t/MusicBrainz/Server/Data/Role/GIDEntityCache.pm [hunk 4]
```diff
@@ -154,7 +164,77 @@ test 'Cache is transactional (MBS-7241)' => sub {
...
+        die $error;
+    }
+
+    $c1->store->delete('artist:recently_invalidated:3');
+
+    # Test 2:
+    #  1. Process A deletes artist 3 (but doesn't commit yet).
+    #  2. Process B fetches artist 3 before the cache deletion, after the
+    #     cache deletion but before process A commits, and finally after
+    #     process A commits.
```
Comment (yvanzo) [inline t/lib/t/MusicBrainz/Server/Data/Role/GIDEntityCache.pm:137]: Maybe make it easier to follow the three sub-tests:

```suggestion
    #  2. Process B fetches artist 3:
    #     a. before the cache deletion,
    #     b. after the cache deletion but before process A commits,
    #     c. and finally after process A commits.
```

File: t/lib/t/MusicBrainz/Server/Data/Role/GIDEntityCache.pm [hunk 5]
```diff
@@ -189,24 +269,25 @@ test 'Cache is transactional (MBS-7241)' => sub {
...
-        SQL
+    try {
+        Sql::run_in_transaction(sub {
+            $c1->model('Artist')->delete(3);
+        }, $c1->sql);
+
+        my $artist = $c1->model('Artist')->get_by_id(3);
+        ok(!defined $artist,
+            'get_by_id returns undef after ' .
+            'database deletion commits, and after cache deletion');
```
Comment (yvanzo) [inline t/lib/t/MusicBrainz/Server/Data/Role/GIDEntityCache.pm:186]: a

Comment (yvanzo) [reply t/lib/t/MusicBrainz/Server/Data/Role/GIDEntityCache.pm:186]: To follow the above suggestion, and since database deletion commits after cache deletion anyway:
```suggestion
        ok(!defined $artist,
            '(c.) get_by_id returns undef after ' .
            'database deletion commits');
```

Similarly, (a.) and (b.) could be inserted into their respective test message.

**Review by yvanzo — COMMENTED** [review]
Thanks for the added tests using nested transactions, because race conditions would be very difficult to reproduce otherwise!

Found minor oversights in text mostly but still have one question:

Any reason not to mark the artist as _recently invalidated_ in cache even before updating it in the database? That would further reduce the chances to have a cached stale artist.

File: lib/MusicBrainz/Server/Data/Artist.pm [hunk 6]
```diff
@@ -339,7 +339,13 @@ sub delete
...
-sub merge
+sub merge {
+    my ($self, $new_id, @old_ids) = @_;
+
+    $self->merge_with_opts($new_id, \@old_ids);
+}
+
+sub merge_with_opts
 {
     my ($self, $new_id, $old_ids, %opts) = @_;
```
[the remark this answers is not in the document]
Comment (yvanzo) [reply lib/MusicBrainz/Server/Data/Artist.pm:350]: Never mind, you clarified it in the meantime in commit https://github.com/metabrainz/musicbrainz-server/pull/3130/commits/c4ae12b1f3dedb80be94e525f6c1c99cd583f48e.

File: lib/MusicBrainz/Server/Data/Role/EntityCache.pm [hunk 7]
```diff
@@ -25,10 +25,16 @@ sub _cache_id {
     }
 }
 
+has _cache_prefix => (
+    is => 'ro',
+    isa => 'Str',
+    default => sub { shift->_type . ':' },
```
[the remark this answers is not in the document]
Comment (yvanzo) [reply lib/MusicBrainz/Server/Data/Role/EntityCache.pm:64]: This is now commit 607213200b76324055ee3fcec97143dc81db38f2.

File: t/lib/t/MusicBrainz/Server/Data/Artist.pm [hunk 8]
```diff
@@ -295,15 +295,7 @@ $sql->commit;
...
-    my $opts = DBDefs->CACHE_MANAGER_OPTIONS;
-    $opts->{profiles}{external}{options}{namespace} = 'mbtest:';
-
-    my $c = $test->c->meta->clone_object(
-        $test->c,
-        cache_manager => MusicBrainz::Server::CacheManager->new(%$opts),
-        models => {}, # Need to reload models to use this new $c
-    );
-
+    my $c = $test->cache_aware_c;
```
[the remark this answers is not in the document]
Comment (yvanzo) [reply t/lib/t/MusicBrainz/Server/Data/Artist.pm]: This is now 519f1a4f6948f828e8d359cc4348aee11e0ad2bc.

File: lib/DBDefs/Default.pm [hunk 9]
```diff
@@ -192,6 +192,26 @@ sub GOOGLE_CUSTOM_SEARCH { '' }
...
+# and the store (for sessions), as these may point to different Redis
+# instances.
+#
+# `REDIS_TEST_DATABASE` indicates the minimum DB number used when running
+# tests. This is used for both the cache and the store. Note that the tests
+# may use additional DB numbers greater than or equal to
+# `REDIS_TEST_DATABASE`, and that all production DB numbers should be less
+# than it. All test databases will be completely erased on each test run.
+sub REDIS_CACHE_DATABASE { 0 }
+sub REDIS_STORE_DATABASE { 0 }
```
[the remark this answers is not in the document]
Comment (yvanzo) [reply lib/DBDefs/Default.pm]: That was in commit https://github.com/metabrainz/musicbrainz-server/commit/4ce0c3563e706f90bd02873a464407bb16867a37 which has been superseded with commit c34c72214b08bfe8234b9330fa470ebf2aa29a33.

Is it needed to specify `database => 0` if it is the default value in Redis?

**Comment by mwiencek** [comment]
> Any reason not to mark the artist as recently invalidated in cache even before updating it in the database? That would further reduce the chances to have a cached stale artist.

IIUC this is already the case: in `_delete_from_cache`, `set_multi(@invalidated_flags)` (paraphrasing) is already called before `delete_multi(@ids)`; and `_delete_from_cache` itself is called before the database transaction which updates/deletes the entity commits.

File: t/lib/t/Context.pm [hunk 10]
```diff
@@ -33,6 +33,7 @@ sub _build_cache_aware_context {
         $test->c,
         cache_manager => MusicBrainz::Server::CacheManager->new(%$opts),
         models => {}, # Need to reload models to use this new $c
+        fresh_connector => 1,
```
[the remark this answers is not in the document]
Comment (mwiencek) [reply t/lib/t/Context.pm]: `t::Context` is only used by the tests, so it will increase the number of connections used by the tests. :)

File: lib/DBDefs/Default.pm (hunk 9 above)
[the remark this answers is not in the document]
Comment (mwiencek) [reply lib/DBDefs/Default.pm]: > Do both cache and store have database 0 because these are hold by different Redis instances? What about mirror which uses only one Redis instance? What would be the inconvenience of setting 1 to store?

 * The cache and store don't have to be separate in order to use 0 for both: we also use different cache prefixes to avoid collisions, so they can both live in database 0 in the same instance.
 * I otherwise assume 0 is only used for both because it's the default.
 * I'm not sure it would necessarily be an inconvenience, though when connecting with `redis-cli` you'll have to pass the `-n` flag to select the appropriate database number.

> That was in commit https://github.com/metabrainz/musicbrainz-server/commit/4ce0c3563e706f90bd02873a464407bb16867a37 which has been superseded with commit https://github.com/metabrainz/musicbrainz-server/commit/c34c72214b08bfe8234b9330fa470ebf2aa29a33.

I removed the `REDIS_CACHE_DATABASE` and `REDIS_STORE_DATABASE` flags because existing DBDefs files (which are typically based on DBDefs.pm.sample) would not even reference them at all.

> Is it needed to specify database => 0 if it is the default value in Redis?

It shouldn't be, though I added it as a form of documentation (to indicate that it's a property which can be set).

[the remark this answers is not in the document]
Comment (mwiencek) [reply lib/DBDefs/Default.pm]: I will add, though, that storing the sessions in DB1 would have a benefit in that you could easily clear the cache without dropping login sessions. (That's only a problem if you're sharing an instance.)

File: lib/DBDefs/Default.pm [hunk 11]
```diff
@@ -248,21 +270,12 @@ sub SESSION_STORE { 'Session::Store::MusicBrainz' }
...
-# redis server.
-
 sub DATASTORE_REDIS_ARGS {
     my $self = shift;
     return {
-        database => 0,
+        database => $self->REDIS_STORE_DATABASE,
         namespace => $self->CACHE_NAMESPACE,
         server => '127.0.0.1:6379',
-        test_database => 1,
```
[the remark this answers is not in the document]
Comment (mwiencek) [reply lib/DBDefs/Default.pm:265]: It was used, but only by a couple of tests under t/lib/t/MusicBrainz/DataStore/. The new DBDefs setting `REDIS_TEST_DATABASE` is intended to be used for all Redis tests (whether for the cache or store).

File: docker/musicbrainz-tests/DBDefs.pm [hunk 12]
```diff
@@ -72,6 +72,8 @@ MusicBrainz::Server::DatabaseConnectionFactory->register_databases(
     },
 );
 
+sub REDIS_TEST_DATABASE { 0 }
```
[the remark this answers is not in the document]
Comment (mwiencek) [reply docker/musicbrainz-tests/DBDefs.pm:75]: I don't remember, so I'll try removing it and see if anything breaks.

File: t/lib/t/MusicBrainz/Server/Data/Role/GIDEntityCache.pm [hunk 13]
```diff
@@ -154,7 +164,77 @@ test 'Cache is transactional (MBS-7241)' => sub {
...
+        # our stale cache addition.
+        $c1->model('Artist')->get_by_gid($artist_gid);
+        $c1->sql->commit;
+
+        ok(!$c1->cache->exists('artist:3'),
+           'artist is not cached after a concurrent update');
+
+        $c1->sql->begin;
+        my $artist = $c1->model('Artist')->get_by_gid($artist_gid);
+        is($artist->name, 'COOL', 'get_by_gid returns the latest changes');
```
[the remark this answers is not in the document]
Comment (mwiencek) [reply t/lib/t/MusicBrainz/Server/Data/Role/GIDEntityCache.pm:118]: I think it was just my reaction when the tests finally started passing. :)

File: t/lib/t/MusicBrainz/Server/Data/Role/GIDEntityCache.pm [hunk 14]
```diff
@@ -154,7 +164,77 @@ test 'Cache is transactional (MBS-7241)' => sub {
...
+            $c2->model('Artist')->update(3, {name => 'COOL'});
+            $c2->sql->commit;
+        }
+        $set_multi->(@_);
+    };
+
+    try {
+        $c1->sql->begin;
+        # Attempt to cache this artist. A concurrent update in another
+        # DB process will update the artist before this one commits
```
[the remark this answers is not in the document]
Comment (mwiencek) [reply t/lib/t/MusicBrainz/Server/Data/Role/GIDEntityCache.pm:107]: Both are true, but I'll use your wording because it's clearer and more relevant to what's being tested.

File: lib/MusicBrainz/Server/Data/Role/EntityCache.pm [hunk 15]
```diff
@@ -3,34 +3,73 @@ package MusicBrainz::Server::Data::Role::EntityCache;
...
+The `recently_invalidated` keys have their TTLs set to the "max request time"
+determined by `DBDefs::DETERMINE_MAX_REQUEST_TIME`. The idea behind this is
+that if the request times out, any database handles used in the request will
+be closed and all transactions will be rolled back. In case the max request
+time is undefined or 0 (indicating no limit), we set the TTL to
+`$RECENTLY_INVALIDATED_TTL`, which defaults to 10 minutes.
+
+The process described above is only performed for `RT_MASTER` and
+`RT_STANDALONE`. `RT_MIRROR` should not have any modifications to the
+database outside of replication.
```
[the remark this answers is not in the document]
Comment (mwiencek) [reply lib/MusicBrainz/Server/Data/Role/EntityCache.pm:47]: My thinking was that those modifications only happen at the database level, via the `ProcessReplicationChanges` script, not via our Perl data models. However, I think you have a point in that the PruneCache script still calls `_delete_from_cache`, so the `recently_invalidated` keys can serve a purpose on mirrors. I'll just change it to run regardless of mirror status.

File: t/lib/t/MusicBrainz/Server/Data/Role/GIDEntityCache.pm [hunk 16]
```diff
@@ -130,22 +142,20 @@ is ( $test->c->cache->_orig->set_called, 2 );
 };
 
 test 'Cache is transactional (MBS-7241)' => sub {
```
[the remark this answers is not in the document]
Comment (mwiencek) [reply t/lib/t/MusicBrainz/Server/Data/Role/GIDEntityCache.pm:57]: I think MBS-7241 makes sense because it concerns issues that occur when the cache is not transactional, whereas MBS-13449 is about performance issues with the implementation of MBS-7241.

File: t/lib/t/MusicBrainz/Server/Data/Artist.pm (hunk 8 above)
[the remark this answers is not in the document]
Comment (mwiencek) [reply t/lib/t/MusicBrainz/Server/Data/Artist.pm]: That was just refactoring (I amended the commit message to indicate this).

**Comment by mwiencek** [comment]
@brainzbot test this please

**Review by yvanzo — APPROVED** [review]
Reviewed replies, pushed changes and commit messages, all good to go! :ship:
